### DISCUSSION TITLE
LoadTexturesAsyncでクラッシュしてしまう

### DISCUSSION THREAD

**Discussion by kayacKou** [pixiv/vroid-sdk-developers#42] [General]
お世話になっております。

VroidSDKのサンプル(CharacterModelExample)を実行した際に、
Android端末、iOS端末でクラッシュしてしまう現象が発生しております。

SDKバージョン0.1.0から発生しており、
SDKバージョン0.0.21では正常にモデルが表示されておりました。

LoadTexturesAsyncに差しかかったところでクラッシュしているようなのですが、
解決法や原因をご存知であればご教授いただけると幸いです。

使用ツール情報
・UnityVersion 2019.4.8f1
・AndroidStudio 2021.1.1 Patch 1

使用しているモデル
Newサンプルモデルの3体

**Comment by kayacKou** [comment]
調査の結果、モデルのテクスチャをLoadImageで生成している箇所でメモリ不足でのクラッシュが起きているようでした。
こちらについて有効な方法、または解決策などあればご教授いただけると幸いです。

**Comment by nyamadan** [comment]
0.0.21ですとメモリ不足が発生していないというところが気になります。
0.1.0では基礎設計の見直しのほか、依存ライブラリであるUniVRMのバージョンアップを行っております。

これらが影響している可能性はありますが具体的に何が原因か今のところ分かりません。

調査をいたしますので、ご不便をおかけいたしますが今しばらくお待ちください。

**Comment by kayacKou** [comment]
お世話になっております。
ご対応ありがとうございます

LoadTexturesAsyncでのメモリ不足でクラッシュしているのではという結果に至った経緯ですが、
内部で行われているTexture2D ::LoadImageをコメントアウトすることでクラッシュは回避できるためです。

また効果はありませんでしたが、
以下はこちらで試したことです
・ヒープサイズを先に拡張しておく
・Load Imageの第２引数をTrueにする

**Comment by nyamadan** [comment]
情報共有ありがとうございます。

調査を進めておりますが今のところ0.0.21と0.1.0でクラッシュにつながるようなメモリアロケーションの差は今のところ見つけられておりません。

メモリ不足が原因であるとすれば表示できるモデル・表示できないモデルがあると思うのですがそういったことはありますか？

**Reply by nyamadan** [reply]
たとえばメモリ不足が原因であれば小さいテクスチャであれば表示できると考えています。

**Reply by kayacKou** [reply]
ご連絡遅くなり、申し訳ございません
クラッシュするしないモデルはそれぞれ存在しております。

Newサンプルモデルの3体はクラッシュしますが、
以下のURLのモデルはクラッシュすることなく読み込めております。
https://hub.vroid.com/characters/2603882572555787982/models/8648523590158739595
https://hub.vroid.com/characters/54798681159054334/models/6727878805689695209
https://hub.vroid.com/characters/5389258699935723913/models/7119927224519623284

**Reply by nyamadan** [reply]
お待たせしました。以下の条件で動作確認しましたところクラッシュは発生しませんでした。

 * Pixel 5
 * Android Studio 2021.1
 * Unity 2019.4.31

機種依存のものではないかと考えています。
クラッシュが発生したAndroid / iOSの機種を教えていただけますか？

**Reply by kayacKou** [reply]
ご連絡ありがとうございます。

こちらでの確認端末は
SHARP AQUOS SHG05になります
https://jp.sharp/products/aquos-sense6/a/

RAMのスペックにだいぶ差があるようなので、
もしよろしければVroidSDKを使用する際の推奨スペックなどもご教授いただけると幸いです。
（こちらで確認したところ見当たらなかったのですがどこかに記載があれば申し訳ないです

クラッシュ時のログも添付しましたので、何かのお役に立てば幸いです。
[SDK_crash.txt](https://github.com/pixiv/vroid-sdk-developers/files/8504122/SDK_crash.txt)

**Reply by kayacKou** [reply]
Mi Note10 Lite
メモリ（RAM）6.00GB

こちらの端末でもクラッシュするようです。

**Reply by nyamadan** [reply]
機種とログの共有ありがとうございます。
調査を進めておりますが共有していただいた機種が手元になく類似機種で調査中ですが、今のところクラッシュを確認できておりません。
改めてAndroidの実機でメモリ使用量をプロファイラで確認してみたところやはり0.0.21と0.1,0でメモリの使用量に違いがみられませんでした。

ほかに心当たりとしましては0.1.0からUniVRM 0.89.0を使用するようになりました。
問題の切り分けを行いたいので、可能であればVRoid SDKを使用せずUniVRM 0.89.0単体でサンプルモデルが表示できるかを確認していただけますと大変助かります。

> RAMのスペックにだいぶ差があるようなので、
> もしよろしければVroidSDKを使用する際の推奨スペックなどもご教授いただけると幸いです。

VRoid SDKはその資質上、組み込み先のアプリや読み込むモデルによって必要となってくるスペックが異なります。
そのため、推奨スペックを定義することが難しく、今のところ推奨スペックを定義しておりません。

引き続き調査をいたしますので何か分かりましたらこちらで報告いたします。

**Reply by kayacKou** [reply]
お世話になっております。
現在UniVRM単体で表示の確認をしております。
シーン内に配置したサンプルのVRMモデルは正常に読み込まれます
バイナリからのランタイムロードをVRoidSDKとできるだけ合わせたいのですが、
参考になりそうなファイル名や処理流れなどをご存じでしたらご教授いただけると幸いです。

**Reply by kayacKou** [reply]
お世話になっております
こちらで確認したところ、特に変わらずクラッシュしているようでした。
確認用のソースを添付してりますので、
何か問題などありましたらご教授いただけると幸いです。

生成されたPrefabをシーン内に直接配置した場合などは正しくモデルが読み込まれております

[UniVRMSample.txt](https://github.com/pixiv/vroid-sdk-developers/files/8559293/UniVRMSample.txt)

**Reply by nyamadan** [reply]
> 現在UniVRM単体で表示の確認をしております。
> シーン内に配置したサンプルのVRMモデルは正常に読み込まれます

協力ありがとうございます。UniVRMのバージョンアップが原因かとも思ったのですがそうとも言い切れなさそうですね。
CharacterModelExampleでは `ModelLoader` の `LoadVRMFromBinary` を使用して実装しております。
こちらを参考に動作確認いただければと存じます。

また、メモリ搭載量が比較的すくない4GBの[SHG06](https://www.au.com/mobile/product/smartphone/shg06/)で動作確認をしたところ特にクラッシュなく表示できていることを確認しました。

手元でクラッシュの再現がまだできておらず、もしメモリのほかにクラッシュの心当たりがございましたら共有いただけますと幸いです。

**Reply by nyamadan** [reply]
サンプルコードを拝見いたしました。

こちら、39行目の `SynchronizationContext.Current` がメインスレッドの `SynchronizationContext` である必要があります。

[こちらのページ](https://qiita.com/toRisouP/items/a2c1bb1b0c4f73366bc6)が参考になると思います。

**Reply by kayacKou** [reply]
お世話になっております。
参考のページのご共有ありあとうございます。

いただいた資料を元に修正をしてみましたが特に変わらずクラッシュするようでした。
メモリの他に関係ありそうなところですと
・Andorid12
・Target API30
・Texture Compression　don't override
あたりでしょうか

また何かわかりましたら、ご共有いただけると幸いです。

[UniVRMSample.txt](https://github.com/pixiv/vroid-sdk-developers/files/8570917/UniVRMSample.txt)

**Reply by nyamadan** [reply]
情報共有ありがとうございます。

こちらのサンプルでも同様に `LoadTexturesAsync` でクラッシュしますでしょうか？
同様のエラーであれば共有していただいたコードに絞って調査してみようと考えています。

**Reply by kayacKou** [reply]
お世話になっております。
ご対応ありがとうございます。
そちらも同様にLoadTexturesAsync内でクラッシュしております。

**Reply by nyamadan** [reply]
お待たせいたしました。

UniVRM 0.97で次の修正がはいったようです。
 https://github.com/vrm-c/UniVRM/pull/1581

VRoid SDK 0.1.0ではUniVRM 0.89.0を使用しており、こちらの修正は含まれておりません。
たびたび恐れ入りますが最新版のUniVRMにて再度お試しいただけますでしょうか？

**Reply by kurino-bond** [reply]
@nyamadan 
初めまして。
@kayacKou と同じプロジェクトで開発対応している者です。

UniVRMの情報共有ありがとうございます。
メモリ確保についての修正内容のため、本件のクラッシュと関係がありそうです。

UniVRM 0.98（最新）で確認したところ、VroidSdk内の処理で以下エラーが発生しました。
（0.97でも確認しましたが、同様です。）
こちらの解決方法についてお心当たりありますでしょうか。

```
Assets/VRoidSDK/Plugins/Pixiv/VroidSdk/Api/Legacy/HubModelDeserializer.cs(143,29): error CS7036: There is no argument given that corresponds to the required formal parameter 'awaitCaller' of 'ImporterContext.LoadAsync(IAwaitCaller, Func<string, IDisposable>)'

Assets/VRoidSDK/Plugins/Pixiv/VroidSdk/ModelLoader.cs(120,29): error CS7036: There is no argument given that corresponds to the required formal parameter 'awaitCaller' of 'ImporterContext.LoadAsync(IAwaitCaller, Func<string, IDisposable>)'

Assets/VRoidSDK/Plugins/Pixiv/VroidSdk/ModelLoader.cs(156,29): error CS7036: There is no argument given that corresponds to the required formal parameter 'awaitCaller' of 'ImporterContext.LoadAsync(IAwaitCaller, Func<string, IDisposable>)'

```

**Reply by kurino-bond** [reply]
情報追記となります。おそらく下記が原因と思われます。
・ImporterContext.LoadAsyncのIAwaitCallerがnull非許容になっていて、ModelLoader.csでエラーが起きている。
・UniVRMの方もVrmUtility.csでエラーが起きる可能性がある（LoadAsyncでIAwaitCaller=nullを投げていて、Exceptionが返される可能性あり）

**Reply by nyamadan** [reply]
初めまして、 [nyamadan](https://github.com/pixiv/vroid-sdk-developers/discussions/3#discussioncomment-1877844) です。
ご不便をおかけして申し訳ございません。

> UniVRM 0.98（最新）で確認したところ、VroidSdk内の処理で以下エラーが発生しました。

[VRoid SDKを使用しないこちらのサンプル](https://github.com/pixiv/vroid-sdk-developers/discussions/42#discussioncomment-2645381)をUniVRM 0.98で動作させた場合も同様にエラーが発生しますでしょうか？

VRoid SDKに組み込んだ場合については心当たりがございません。バージョンアップにより実装がかわったのでしょうか。
調査いたしますので、今しばらくお待ちいただければと思います。

**Reply by kurino-aks** [reply]
@nyamadan 
エラーについては、 IAwaitCallerを継承していたImmediateCallerを
空で作って渡すことで解消することが出来ました。

```
VRMShaders.ImmediateCaller awaitCaller = new VRMShaders.ImmediateCaller();
importerContext.LoadAsync(awaitCaller).ContinueWith((result) =>
```

ただ、相変わらずモデルダウンロードでクラッシュするため、
アップデート後のUniVRM単体で落ちるかどうかを確認中です。

モデル読み込み時のクラッシュ時のメモリ状況も確認しておりますが、1G付近でクラッシュしているようです。
使用端末：Xiaomi mi10 Lite
<img width="640" alt="スクリーンショット 2022-05-11 16 05 54" src="https://user-images.githubusercontent.com/104487566/167796473-e70117ea-cfa2-4a82-96f5-f151be81b8a2.png">

**Reply by miura-bond** [reply]
@nyamadan
初めまして。
@kayacKou @kurino-bond と同じプロジェクトで開発対応している者です。

UniVRM 0.98単体でもクラッシュを確認しました。
また、以前は表示されていたSDKバージョン0.0.21に付属していたUniVRM 0.55.0単体でもクラッシュが発生しております。
クラッシュ箇所は以前同様Texture2D.LoadImageなので、UniVRM(UniGLTF)の部分となります。

クラッシュ端末ですが、そもそもUnityの「Development Build」をオンにしただけでもクラッシュが発生するため、端末依存の可能性が高まっております。
また、UniVRM 0.98ではColorSpaceがLinear固定でOpenGLES2.0が非対応となり、
リリース済みのアプリの為に0.89に戻そうかと考えております。

他の方法として読み込み部分をTexture2D.LoadRawTextureDataを利用すると軽くなるということでその方法も考えておりますが、
現状上手くテクスチャが読み込めないのと、そもそもUniVRMのSDKの部分なので難しい所です。
かなり狭まった部分の話にはなってしまいますが、こちら、何か知見はございますでしょうか。

下に改めて環境情報を纏めております。
よろしくお願いいたします。

環境情報
Unity: 2019.4.8f1
Android Studio: 2021.1.1
Gradle: 3.4.0
VRoidSDK: 0.2.0
UniVRM: 0.98

端末
Xiaomi mi10 Lite
MIUI Global 12.5.4
Android 11

**Reply by nyamadan** [reply]
> また、UniVRM 0.98ではColorSpaceがLinear固定でOpenGLES2.0が非対応となり、
> リリース済みのアプリの為に0.89に戻そうかと考えております。

VRoid SDKは最新版のUniVRMの追従する方針のため、今後リリースされるVRoid SDKはUniVRM次第ですが必然的にLinear色空間が必須となりますのでご注意ください。

> 他の方法として読み込み部分をTexture2D.LoadRawTextureDataを利用すると軽くなるということでその方法も考えておりますが、
> 現状上手くテクスチャが読み込めないのと、そもそもUniVRMのSDKの部分なので難しい所です。
> かなり狭まった部分の話にはなってしまいますが、こちら、何か知見はございますでしょうか。

テクスチャロードのメモリ負荷が原因なのであれば呼び出すメソッドを変えれば改善する可能性はあるかもしれません。
ですがVRMにはRGBA32の非圧縮テクスチャが格納されていることがほとんどであると考えられます。
そうなると必然的にメモリの使用量は増えるため、性能が十分でない端末ですとクラッシュしやすいということもあるかもしれません。

お力になれず申し訳ありません。
